New BigQuery Driver

jeff303 opened the pull request:
New BigQuery Driver

Create new `:bigquery-cloud-sdk` driver using the `google-cloud-bigquery` library instead (copied from `:bigquery` driver)

Update new driver and query processor code to use newer Google SDK

Switch test data loading over to use new API

Add optional `project-id` connection property to override the value from the service account JSON

Updating google driver so its libraries are compatible with the newer ones used in BigQuery

Update date bucketing tests to skip `:bigquery-cloud-sdk` (new driver) where `:bigquery` is skipped

jeff303 committed:
New BigQuery Driver

Create new :bigquery-cloud-sdk driver using the google-cloud-bigquery library instead (copied from :bigquery driver)

Update new driver and query processor code to use newer Google SDK

Switch test data loading over to use new API

Add project-id connection property to override the value from the service account JSON

Updating google driver so its libraries are compatible with the newer ones used in BigQuery

Update date bucketing tests to skip :bigquery-cloud-sdk (new driver) where :bigquery is skipped

Update `with-bigquery-fks` to take in the driver, since there are now multiple ones

jeff303 committed:
u/get-id -> u/the-id

jeff303 committed:
Fix typo in kw

Add license overrides for Google libs

camsaul commented on .circleci/config.yml:
```diff
@@ -981,6 +982,12 @@ workflows:
           driver: bigquery
 
+      - test-driver:
+          name: be-tests-bigquery-cloud-sdk-ee
```
why not just update the existing BigQuery driver to use the Cloud SDK? We don't need to make a completely new driver every time the underlying dependencies change

jeff303 replied:
It's more or less a complete rewrite, and the general consensus is this introduces a lot of risk.  So Reza wants it to go through the deprecation cycle.  Besides that, it's dropping support for the non-service-account auth mechanisms entirely (not even work on upgrade), so it's not backward compatible.

jeff303 replied:
This has been rebased and now uses the deprecation logic from #17028.

jeff303 commented:
Final merge should not happen until #17028 is merged, at which point, the existing BigQuery driver can be marked as superseded by the new one.  Also, this PR will be retargeted to 40.

howonlee commented:
I've been using this all clienty-like and I'm going to note here that there's no provision in the DB creation screen front-end to upload the client json. I've just been manually reaching into database model in repl and creating them that way but obviously end user isn't going to do this

jeff303 commented:
> I've been using this all clienty-like and I'm going to note here that there's no provision in the DB creation screen front-end to upload the client json. I've just been manually reaching into database model in repl and creating them that way but obviously end user isn't going to do this
> 
> ![Screenshot from 2021-07-20 14-30-32](https://user-images.githubusercontent.com/1364952/126398024-0d13f077-8ceb-49f1-a8f1-d731c0bd512a.png)

Thanks, Howon.  Will ensure this is addressed before the rebase/merge.

jeff303 committed:
Incorporate deprecation logic for BigQuery

jeff303 committed:
Add Cypress test

viniciusdsmello commented:
Hi @jeff303, I've found this PR by searching for a solution for one of the problems I'm facing with Metabase. I want to use a service account from Project A to connect to a dataset on Project B so that the billing and also quota consumption will go to the first project. The problem is that Metabase doesn't allow me to configure a different project-id with a hyphen.

Do you know if your PR will address it ? I've found on master that this line does a check against bigquery identifiers 

https://github.com/metabase/metabase/blob/92451f958ae2a72930c1d70330a878e2c3585ba7/modules/drivers/bigquery/src/metabase/driver/bigquery/query_processor.clj#L35

viniciusdsmello committed:
Fixes valid-bigquery-identifier regex

Fixes regex that checks if a dataset-id is valid or not. Adds dot and hyphen as allowed characters so datasets from others projects can be passed.

E.g.: project-id.dataset-id or dataset-id

Still needs do change frontend regex

viniciusdsmello commented:
@jeff303 I've tested this change that I'd mentioned and it works. There're changes that still need to be made, for example, the frontend form also does validation on the dataset-id. We're going to use a modified image until this change is released.

I've created a PR with this change: https://github.com/metabase/metabase/pull/17235

viniciusdsmello committed:
Escape punctuation

viniciusdsmello committed:
Fixes valid-bigquery-identifier regex to take into account maximum string lenght

viniciusdsmello committed:
Adds test for documented bigquery dataset name

viniciusdsmello committed:
Fixes test for rejected names

jeff303 committed:
Bring changes from #17235 to new BigQuery driver

jeff303 commented:
> @jeff303 I've tested this change that I'd mentioned and it works. There're changes that still need to be made, for example, the frontend form also does validation on the dataset-id. We're going to use a modified image until this change is released.
> 
> I've created a PR with this change: #17235

I incorporated your commits into this PR (to get the full test suite), and also adapted them to the new BigQuery driver (along with the one that will become legacy).  Do you mind checking out the uberjar [here](https://app.circleci.com/pipelines/github/metabase/metabase/19415/workflows/186a3230-9e87-431b-a435-516ac1809fe6/jobs/798855/artifacts) and seeing if it works for you?

You had left a comment [here](https://github.com/metabase/metabase/pull/17235/files#diff-5896a28d56e5da7e01e9b7219c2d890078f777f9f9cb82407693d4ac0cc3bb07R35) about needing a frontend change, but I'm not sure what that's referring to.  Can you elaborate more?

jeff303 committed:
Fix linter warnings

viniciusdsmello commented:
> > @jeff303 I've tested this change that I'd mentioned and it works. There're changes that still need to be made, for example, the frontend form also does validation on the dataset-id. We're going to use a modified image until this change is released.
> > I've created a PR with this change: #17235
> 
> I incorporated your commits into this PR (to get the full test suite), and also adapted them to the new BigQuery driver (along with the one that will become legacy). Do you mind checking out the uberjar [here](https://app.circleci.com/pipelines/github/metabase/metabase/19415/workflows/186a3230-9e87-431b-a435-516ac1809fe6/jobs/798855/artifacts) and seeing if it works for you?
> 
> You had left a comment [here](https://github.com/metabase/metabase/pull/17235/files#diff-5896a28d56e5da7e01e9b7219c2d890078f777f9f9cb82407693d4ac0cc3bb07R35) about needing a frontend change, but I'm not sure what that's referring to. Can you elaborate more?

Hi @jeff303, I don't know exactly if it's related to the Frontend, but one thing is, after changing the regex I was able to query a dataset from another project (I had to change the dataset location from Metabase's database), but when I try to add a new database/bigquery dataset I receive an error and It doesn't allow me to save.

```Not found: Dataset service_account_project:external_project.dataset```

Adding a new field called billing_project (defaults to service account project) might fix it.

jeff303 committed:
Switch bigquery-cloud-sdk over to deps.edn

jeff303 committed:
Add entry to modules/drivers/deps.edn

jeff303 committed:
Add entries for :drivers-dev and :eastwood

jeff303 committed:
Fix import ordering

jeff303 commented:
> Hi @jeff303, I don't know exactly if it's related to the Frontend, but one thing is, after changing the regex I was able to query a dataset from another project (I had to change the dataset location from Metabase's database), but when I try to add a new database/bigquery dataset I receive an error and It doesn't allow me to save.
> 
> `Not found: Dataset service_account_project:external_project.dataset`
> 
> Adding a new field called billing_project (defaults to service account project) might fix it.

You tried setting the `Project ID (override)` field?

viniciusdsmello commented:
> You tried setting the `Project ID (override)` field?

I didn't try it, when I configure a service account through the UI, there is no field to set the Project ID:

![image](https://user-images.githubusercontent.com/6565443/128028983-41e82736-0a0f-4058-8e82-2ddf31842efc.png)

The Project ID is inferred from the service account, so if I want to access a dataset from a different project I need to pass it on dataset id:

![image](https://user-images.githubusercontent.com/6565443/128037060-55eb8d45-8145-4311-93d9-b6925770b78f.png)

jeff303 commented:
@viniciusdsmello , it appears that you are still showing screenshots using the older driver.  To be very clear, the only way you will see the newer driver (as of this moment) is to use the uberjar built as part of **this PR** (which has not even been merged into nightly/`master` branch).  It will look something like this.

<img width="1069" alt="Screen Shot 2021-08-03 at 10 22 07 AM" src="https://user-images.githubusercontent.com/3521562/128041867-fba8df47-e58f-4ac8-9b59-49f8b53df7e6.png">

viniciusdsmello commented:
Hi @jeff303 , I tried this uberjar, and It seems that it is not working correctly, at least not as I expected it to work. I tried to set "Project ID (override)" to the external Project where the dataset is located, but the query created is not using this info.  ​I've checked bigquery, and the external project isn't being used set.

I also tried to set the project-id with the dataset-id, but doing this, I receive an error.

jeff303 commented:
> Hi @jeff303 , I tried this uberjar, and It seems that it is not working correctly, at least not as I expected it to work. I tried to set "Project ID (override)" to the external Project where the dataset is located, but the query created is not using this info. ​I've checked bigquery, and the external project isn't being used set.
> 
> I also tried to set the project-id with the dataset-id, but doing this, I receive an error.

Are you able to post the native query that is generated, and what you are expecting it to be?

jeff303 committed:
Stop overwriting project-id with value from credentials

viniciusdsmello commented:
> Are you able to post the native query that is generated, and what you are expecting it to be?

Hi @jeff303, sorry for the late reply. 

There are two project ids. One is related to the service account (billing project) and the other to the dataset being queried (dataset location).

The connection to BigQuery must be created with the service account's project-id and must allow query data from another dataset. So, let me use **project-svc-account** and **project-dataset** as an example, and suppose there is a table my-table and a dataset my-dataset in project-dataset.

The connection must be created to project-svc-account, and the query must be something like

In a connection to project-svc-account: SELECT * FROM `project-dataset.my-dataset.my-table`

In this way, the billing and quota consumption will go to project-svc-account and the project-dataset won't be charged.

In Python, for example, I just need to do this:

```
from google.cloud import bigquery
client = bigquery.Client(project='project-svc-account')
query = 'SELECT * FROM `project-dataset.my-dataset.my-table`'

query_job = client.query(query)
....
```

jeff303 commented:
> > Are you able to post the native query that is generated, and what you are expecting it to be?
> 
> Hi @jeff303, sorry for the late reply.
> 
> There are two project ids. One is related to the service account (billing project) and the other to the dataset being queried (dataset location).
> 
> The connection to BigQuery must be created with the service account's project-id and must allow query data from another dataset. So, let me use **project-svc-account** and **project-dataset** as an example, and suppose there is a table my-table and a dataset my-dataset in project-dataset.
> 
> The connection must be created to project-svc-account, and the query must be something like
> 
> In a connection to project-svc-account: SELECT * FROM `project-dataset.my-dataset.my-table`
> 
> In this way, the billing and quota consumption will go to project-svc-account and the project-dataset won't be charged.
> 
> In Python, for example, I just need to do this:
> 
> ```
> from google.cloud import bigquery
> client = bigquery.Client(project='project-svc-account')
> query = 'SELECT * FROM `project-dataset.my-dataset.my-table`'
> 
> query_job = client.query(query)
> ....
> ```

Thanks for the details here.  I'm still working internally to get multiple project IDs set up for testing.  In the meantime, there was definitely a bug in the last version you tested where the `project-id` (i.e. the override config field) was always being overwritten with the one associated with the service account.  So in other words, it was never taking effect.  I pushed a commit to fix that particular bug, but still, when I try to test out with public BigQuery data sets, I'm getting an error (albeit a different one: like: `Access Denied: Project bigquery-public-data: User does not have bigquery.jobs.create permission in project bigquery-public-data`).  Perhaps if you have multiple projects with these permissions set up correctly, it would work now.

jeff303 committed:
Fix imports more

viniciusdsmello commented:
> Thanks for the details here. I'm still working internally to get multiple project IDs set up for testing. In the meantime, there was definitely a bug in the last version you tested where the `project-id` (i.e. the override config field) was always being overwritten with the one associated with the service account. So in other words, it was never taking effect. I pushed a commit to fix that particular bug, but still, when I try to test out with public BigQuery data sets, I'm getting an error (albeit a different one: like: `Access Denied: Project bigquery-public-data: User does not have bigquery.jobs.create permission in project bigquery-public-data`). Perhaps if you have multiple projects with these permissions set up correctly, it would work now.

There is the problem, the Job must not be created within the external project. In BigQuery we can set who will be charged for the data access, in this case, the project where the job was created is the one that will be billed. **So, if the service account's project is being overwritten, the error will occur for every public dataset, since the job will be created in the new project set.**

If I want to access a public dataset, I need to create a Job in my project, otherwise, the bills would go to the project hosting the data. For it to work, the service account must have roles/bigquery.jobUser role, and in the external project roles/bigquery.dataViewer.

Maybe adopting two fields, billing-project-id and project-id must fix it. The first one is where the Job will be created and the second one is where the dataset lives.

flamber commented:
Sounds like #2273

jeff303 commented:
OK, I figured out what's going on with the `project-id` stuff.  Stay tuned as I fix up all the tests and get the commit added.

jeff303 committed:
Fix support for separate project-id

Rework the `database->client` function in the driver, since it now doesn't need to do anything with project-id

Update the query processor to set the project-id as part of qualified dataset names, if set, falling back to the previous behavior if not

Adding test to confirm that overriding project-id for a public BQ project works (sync and query)

Fixing a bunch of whitespace alignment errors in tests

jeff303 committed:
Fix fn order

dpsutton commented on modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj:
```diff
@@ -0,0 +1,312 @@
+
+(defn- ^ServiceAccountCredentials database->service-account-credential
+  "Returns a `ServiceAccountCredentials` (not scoped) for the given DB, from its service account JSON."
+  ^ServiceAccountCredentials [{{:keys [^String service-account-json]} :details
```
nit seems you have the typehint twice.

jeff303 replied:
Oops, removed the one before the symbol name, as per: https://clojure.org/reference/java_interop#typehints

dpsutton commented on modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj:
```diff
@@ -0,0 +1,312 @@
+(defmethod driver/describe-database :bigquery-cloud-sdk
+  [_ database]
+  ;; first page through all the 50-table pages until we stop getting "next page tokens"
+  (let [tables (list-tables database) #_(loop [tables [], ^TableList table-list (list-tables database)]
```
why is the commented out code still here?

jeff303 replied:
Oops, removed.

dpsutton commented on modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj:
```diff
@@ -0,0 +1,312 @@
+  [_ details-map]
+  ;; check whether we can connect by seeing whether we have at least one Table in the iterator
+  (let [^Iterable table-iter (list-tables {:details details-map})]
+    (.. table-iter iterator hasNext)))
```
can this distinguish between connecting to a dataset that has no tables in it versus unable to connect to a dataset?

jeff303 replied:
Yes; the latter case will throw an exception (earlier on, actually, in the `database->client` call).

dpsutton replied:
but won't `hasNext` return false on a dataset with no tables, causing `driver/can-connect?` to return false to a dataset it can connect to which has no tables?

dpsutton replied:
```clojure
(.. (list-tables {:details {:project-id "metabase-testing-311615"
                            :service-account-json "redacted"
                            :dataset-id "empty"
                            :engine :bigquery-cloud-sdk}})
    iterator hasNext)
```

this returns false, even though it connects just fine and does not throw an error. It was a new dataset i created `bq mk empty` that has no tables in it. It seems like we would more likely want something like

`(try (list-tables db) true (catch Exception e false))` if list-tables will throw if it cannot connect.

jeff303 replied:
Ah I see your point.  Yes, I will fix that.

jeff303 replied:
OK, this has been fixed now.  `can-connect?` will instead simply check for the presence of the given `dataset-id` (not even looking at tables).  Added a test while I'm at it.

dpsutton commented on modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj:
```diff
@@ -0,0 +1,312 @@
+  ([client :- BigQuery, project-id :- (s/maybe su/NonBlankString), dataset-id :- su/NonBlankString, table-id :- su/NonBlankString]
+   (if project-id
+     (.getTable client (TableId/of project-id dataset-id table-id) (make-array BigQuery$TableOption 0))
+     (.getTable client dataset-id table-id (make-array BigQuery$TableOption 0)))))
```
above you used `u/varargs` for this purpose. Should we be consistent? I don't care which we choose. Also a big fan of just `(def table-options ...)` and reusing that

jeff303 replied:
Yes, that's better.  Also replaced one other `make-array` instance with `u/varargs` for consistency.

dpsutton commented on modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj:
```diff
@@ -0,0 +1,312 @@
+    "NUMERIC"   :type/Decimal
+    :type/*))
+
+#_(s/defn ^:private table-schema->metabase-field-info-old
```
should this be removed?

jeff303 replied:
Yes.

jeff303 committed:
PR feedback

jeff303 committed:
Fixing can-connect? and adding test

jeff303 committed:
Add missing properties to manifest

dpsutton commented on modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/params.clj:
```diff
@@ -0,0 +1,96 @@
+            [metabase.util.date-2 :as u.date])
+  (:import [com.google.cloud.bigquery QueryJobConfiguration$Builder QueryParameterValue StandardSQLTypeName]))
+
+#_(defn- param-type-old ^QueryParameterType [^String type-name]
```
what's the purpose of these three commented out forms?

jeff303 replied:
Ugh, more dead code I forgot to remove.  Should be fixed in newest commit.

dpsutton commented on modules/drivers/google/project.clj:
```diff
@@ -0,0 +1,28 @@
+(defproject metabase/google-driver "1.0.0-SNAPSHOT-1.30.7"
```
I'm surprised to see a `project.clj` file here. Does that play nice with the new `prep` steps?

jeff303 replied:
Yeah this was unintentional.  This PR has drug on so long that the `deps.edn` switchover happened in the middle of it, and somehow this file found its way back in.  Probably a borked merge conflict resolution.  Removed now.

dpsutton reviewed (approved):
I'm surprised by two things
- there is a project.clj file in here. That seems like a relic of a past world
- that the current bigquery driver is marked deprecated. I would have thought this one might be marked "alpha" and released, and then the changeover to deprecated on the current one after one release. But i know yall have been thinking through that a lot. So i'm just registering my surprise without having been in those meetings and discussions.

jeff303 committed:
PR feedback

Remove dead code in params.clj and switch to threaded forms

Remove spurious project.clj that found its way into the google driver

jeff303 commented:
> I'm surprised by two things
> 
> * there is a project.clj file in here. That seems like a relic of a past world
> * that the current bigquery driver is marked deprecated. I would have thought this one might be marked "alpha" and released, and then the changeover to deprecated on the current one after one release. But i know yall have been thinking through that a lot. So i'm just registering my surprise without having been in those meetings and discussions.

Removed the relic (good catch).

As for deprecation, check out the `Driver Cutover Strategy` Notion page, which was refined in collaboration with product.  Basically it will normally happen over the course of two releases, although in practice there is nothing forcing us to actually do the next step in the next major release (I imagine that will be a decision made by product team on a release-by-release basis).  It will be a more risky/difficult decision in the case where drivers can't be automatically migrated by us (for instance, if someone is still using the legacy OAuth stuff for BigQuery).
